### ISSUE TITLE
net-mgmt/telegraf: Unable to fetch HAProxy statistics

### ISSUE THREAD

**Issue by ppmathis** [opnsense/plugins#1232]
**Plugin:** net-mgmt/telegraf 1.7.4 with patch 73804ba due to #1147 
**System:** OPNsense 19.1.3-amd64 with LibreSSL 2.8.3

**Description:** Telegraf is unable to fetch HAProxy statistics through `inputs.haproxy`, for which support was added last year in #842. When checking the log file at `/var/log/telegraf/telegraf.log`, it is clearly visible that Telegraf running as the `telegraf` user lacks permission to access the HAProxy statistics:

```
E! [inputs.haproxy]: Error in plugin: Could not connect to socket '/var/run/haproxy.socket': dial unix /var/run/haproxy.socket: connect: permission denied
```

Currently, the HAProxy socket is owned by `root:root` and is only user-writeable, so the Telegraf daemon has no chance of obtaining statistics. Aside from somehow rearranging socket permissions, [adding support for querying the HTTP statistics page](https://github.com/influxdata/telegraf/tree/master/plugins/inputs/haproxy) that HAProxy offers (`Services -> HAProxy -> Settings -> Settings -> Statistics Configuration`) might also be a feasible solution.

**How to reproduce:** Install the HAProxy and Telegraf plugins, activate the Telegraf HAProxy input at `Services -> Telegraf -> Input`, restart Telegraf and check the logs at `/var/log/telegraf/telegraf.log`. The logs will clearly show that fetching HAProxy statistics is currently not possible due to missing socket permissions.

**Comment by fabianfrz** [comment]
@fraenki and @mimugmail Maybe one of you needs to to change the permissions to make that working.

**Comment by fichtner** [comment]
I know it's a long shot, but it would make sense to change ownership to "proxy" group and make it group-readable (if not already).

**Comment by mimugmail** [comment]
Shouldn't this be enough?

`srwxr-xr-x  1 root  wheel  0 Mar  7 10:26 /var/run/haproxy.socket`

**Comment by mimugmail** [comment]
Ah, it seems you have to enable remote statistics to get the correct settings on the socket.

**Comment by ppmathis** [comment]
@mimugmail I am unable to get correct permissions on the socket, no matter if remote statistics are disabled, enabled or enabled with a dummy listener on another port. They're always stuck on these values, even after restarting the service several times:

```
srwxr-x---  1 root  0  0 Mar  9 11:53 /var/run/haproxy.socket
```

Additionally, the [Telegraf HAProxy input plugin](https://github.com/influxdata/telegraf/blob/master/plugins/inputs/haproxy/haproxy.go#L120-L136) connects to the UNIX socket, writes/sends `show stat\n` and then retrieves the data - wouldn't this require R/W permissions on the socket, which `srwxr-xr-x` wouldn't grant to anybody except root?

Maybe just relying on the HTTP statistics listener would be easier, as it doesn't require messing around with the socket permissions? The HAProxy plugin already supports that, ticking the checkbox `Enable stats` creates a listener per default on `localhost:8822/tcp`, which the Telegraf plugin could query.

**Comment by fraenki** [comment]
> I know it's a long shot since it isn't needed for HAProxy, but it would make sense to change ownership to "proxy" group and make it group-readable (if not already)?

I've updated the socket permissions in https://github.com/opnsense/plugins/pull/1646/commits/8d48327a64989ef418ed32d32383253d4033bd75. Now telegraf seems to be able to get information:

`su -m telegraf -c 'echo "show stat" | nc -U /var/run/haproxy.socket'`

I can't test this with the telegraf service. Can anyone confirm that it's working as intended?

`opnsense-patch -c plugins 8d48327a`
(Need to hit `Apply` in HAProxy afterwards.)

**Comment by fraenki** [comment]
The fix will be available in the upcoming os-haproxy 2.20 – please let me know if it doesn't work as expected.
